Sources/OpenPass/DeviceAuthorizationFlow.swift

```diff
@@ -0,0 +1,211 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+import Foundation
+
+/// A client for Device Authorization, a two-step flow where an input constrained device such as a TV requests a code
+/// and another device inputs this code and provides authorization.
+///
+/// The `DeviceCode` returned in the first step contains a code which should be displayed in a user interface.
+/// The flow then supports polling to check for authorization.
+///
+///     Task {
+///        let flow = OpenPassManager.shared.deviceAuthorizationFlow
+///        do {
+///            // Request a Device Code
+///            let deviceCode = try await flow.fetchDeviceCode()
+///
+///            // Present the code and uri to the user
+///
+///            // Poll for authorization
+///            let tokens = try await flow.fetchAccessTokenPolling(deviceCode: deviceCode)
+///            // do something with tokens
+///        } catch {
+///            // do something with error
+///        }
+///    }
+@MainActor
+public final class DeviceAuthorizationFlow {
+
+    // When polling the token endpoint, it's possible that the response could ask us to "slow down". 
+    // For each slow down response we must add an additional 5 seconds to the original interval.
+    internal var slowDownMultiplier: Int64 = 0
+
+    // The number of additional seconds that should be added to the interval if asked to slow down (the polling).
+    private static let defaultSlowDownFactor: Int64 = 5
+
+    // MARK: - Init
+    
+    private let openPassClient: OpenPassClient
+    private let tokenValidator: IDTokenValidation
+    private let tokensObserver: ((OpenPassTokens) async -> Void)
+    private let dateGenerator: DateGenerator
+    private let clock: Clock
+
+    internal init(
+        openPassClient: OpenPassClient,
+        tokenValidator: IDTokenValidation,
+        tokensObserver: @escaping ((OpenPassTokens) async -> Void),
+        dateGenerator: DateGenerator = .init { Date() },
+        clock: Clock = RealClock()
+    ) {
+        self.openPassClient = openPassClient
+        self.tokenValidator = tokenValidator
+        self.tokensObserver = tokensObserver
+        self.dateGenerator = dateGenerator
+        self.clock = clock
+    }
+
+    // MARK: - Public API
+    
+    /// Start the authorization flow by requesting a Device Code from the API server.
+    /// The returned ``DeviceCode`` is also used with the `fetchAccessToken` and`fetchAccessTokenPolling` methods.
+    /// - Returns: A Device Code representation
+    public func fetchDeviceCode() async throws -> DeviceCode {
```

Comment:
Something to check with @christriddle-thetradedesk, is whether we also want to add support for `loginHint` and `disableLoginHintEditing`. This was added by Riddle later on Android, and we may want to add here (in a separate PR) to be consistent.